crates/bevy_reflect/src/func/return_type.rs

```diff
@@ -0,0 +1,135 @@
+use crate::Reflect;
+
+/// The return type of a [`DynamicFunction`].
+///
+/// [`DynamicFunction`]: crate::func::DynamicFunction
+#[derive(Debug)]
+pub enum Return<'a> {
+    /// The function returns nothing (i.e. it returns `()`).
+    Unit,
+    /// The function returns an owned value.
+    Owned(Box<dyn Reflect>),
+    /// The function returns a reference to a value.
+    Ref(&'a dyn Reflect),
+    /// The function returns a mutable reference to a value.
+    Mut(&'a mut dyn Reflect),
+}
+
+impl<'a> Return<'a> {
+    /// Returns `true` if the return value is [`Self::Unit`].
+    pub fn is_unit(&self) -> bool {
+        matches!(self, Return::Unit)
+    }
+
+    /// Unwraps the return value as an owned value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the return value is not [`Self::Owned`].
+    pub fn unwrap_owned(self) -> Box<dyn Reflect> {
+        match self {
+            Return::Owned(value) => value,
+            _ => panic!("expected owned value"),
+        }
+    }
+
+    /// Unwraps the return value as a reference to a value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the return value is not [`Self::Ref`].
+    pub fn unwrap_ref(self) -> &'a dyn Reflect {
+        match self {
+            Return::Ref(value) => value,
+            _ => panic!("expected reference value"),
+        }
+    }
+
+    /// Unwraps the return value as a mutable reference to a value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the return value is not [`Self::Mut`].
+    pub fn unwrap_mut(self) -> &'a mut dyn Reflect {
+        match self {
+            Return::Mut(value) => value,
+            _ => panic!("expected mutable reference value"),
+        }
+    }
+}
+
+/// A trait for types that can be converted into a [`Return`] value.
```

Comment:
A doc comment here explaining why we don't just use `Into<Return>` would be helpful to teach readers.